src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,172 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines whether <paramref name="embeddedStatement"/> should use braces under a
+        /// <see cref="PreferBracesPreference.WhenMultiline"/> preference due to the presence of braces on one or more
+        /// sibling statements (the "context").
+        /// </summary>
+        private static bool RequiresBracesToMatchContext(SyntaxNode statement, StatementSyntax embeddedStatement)
+        {
+            if (!statement.IsKind(SyntaxKind.IfStatement, SyntaxKind.ElseClause))
+            {
+                // 'if' statements are the only statements that can have multiple embedded statements which are
+                // considered relative to each other.
+                return false;
+            }
+
+            var topLevelContext = GetTopLevelContext(statement);
+            if (AnySiblingUsesBraces(topLevelContext))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the top-most <see cref="IfStatementSyntax"/> for which <paramref name="ifStatementOrElseClause"/> is
+        /// part of the <c>if</c>/<c>else if</c>/<c>else</c> sequence.
+        /// </summary>
+        /// <remarks>
+        /// <para>For the purpose of brace usage analysis, the embedded statments of an <c>if</c>/<c>else if</c>/<c>else</c>
+        /// sequence are considered sibling statements, even though they don't appear as immediate siblings in the
+        /// syntax tree. This method walks up the syntax tree to find the <c>if</c> statement that starts the
+        /// sequence.</para>
+        /// </remarks>
+        private static IfStatementSyntax GetTopLevelContext(SyntaxNode ifStatementOrElseClause)
+        {
+            IfStatementSyntax result;
+            if (ifStatementOrElseClause.IsKind(SyntaxKind.ElseClause))
+            {
+                result = (IfStatementSyntax)ifStatementOrElseClause.Parent;
+            }
+            else
+            {
+                Debug.Assert(ifStatementOrElseClause.IsKind(SyntaxKind.IfStatement));
+                result = (IfStatementSyntax)ifStatementOrElseClause;
+            }
+
+            while (result != null)
+            {
+                if (!result.IsParentKind(SyntaxKind.ElseClause))
+                {
+                    break;
+                }
+
+                result = (IfStatementSyntax)result.Parent.Parent;
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Determines if any embedded statement of an <c>if</c>/<c>else if</c>/<c>else</c> sequence uses braces. Only
+        /// the embedded statements falling <em>immediately</em> under one of these nodes are checked.
+        /// </summary>
+        private static bool AnySiblingUsesBraces(IfStatementSyntax statement)
```

Comment:
based on the docs, and the param type.  I think `AnyIfStatementChainEmbeddedStatementUsesBraces` might be ok.  Another approach (though it allocates) , would be to have a helper like GetEmbeddedStatementsOfIfStatementChain, and hten do .Any on htem to see if they have braces.  However, i can see not wanting to allocate.

Reply:
➡️ Renamed to `AnyPartOfIfSequenceUsesBraces`